fix(table): allow default values to be set

enricoberti opened the pull request:
#429

Closes #429

**Summary**
While we wait for Carbon 10 to be released and referenced, this is a port of their fix: https://github.com/carbon-design-system/carbon/commit/e97a7ed9be2911707066c406544310f6cb76f273

The default value can now be set with `view.toolbar.search.defaultValue`

**Change List (commits, features, bugs, etc)**

- Added a wrapper around TableToolbarSearch

**Acceptance Test (how to verify the PR)**

- Added a new story on ?path=/story/watson-iot-table--with-pre-filled-search

tay1orjones commented on src/components/Table/TableToolbarSearch/_table-toolbar-search.scss:
```diff
@@ -0,0 +1,12 @@
+@import '~carbon-components/scss/globals/scss/vars';
```
This is not needed, it is imported in the local global vars file that you import on the following line
```suggestion

```

tay1orjones commented on src/components/Table/TablePropTypes.js:
```diff
@@ -211,5 +211,5 @@ export const defaultI18NPropTypes = {
 
 export const TableSearchPropTypes = PropTypes.shape({
-  value: PropTypes.string,
+  defaultValue: PropTypes.string,
```
Renaming this prop would be a breaking change. Can we leave it as `value`?

sstone2423 replied:
It looks Carbon is using it in their commited fix. Maybe it could be aliased as `value: defaultValue` ?

tay1orjones commented on src/components/Table/TableToolbarSearch/TableToolbarSearch.jsx:
```diff
@@ -0,0 +1,101 @@
+
+  const onChange = e => {
+    if (onChangeProp) {
+      onChangeProp(e);
```
Is the event `e` not passed if you just directly use onChange from props, instead of renaming it `onChangeProp`?

tay1orjones commented on src/components/Table/TableToolbarSearch/TableToolbarSearch.jsx:
```diff
@@ -0,0 +1,101 @@
+/* istanbul ignore file */
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+/* eslint-disable jsx-a11y/click-events-have-key-events */
```
We should fix the a11y violations, rather than ignore them.

sstone2423 commented on src/components/Table/TableToolbarSearch/TableToolbarSearch.jsx:
```diff
@@ -0,0 +1,101 @@
+const { TableToolbarSearch: CarbonTableToolbarSearch } = DataTable;
+
+const propTypes = {
+  defaultValue: PropTypes.string,
```
Prop comments would be helpful

sstone2423 commented on src/components/Table/TableToolbarSearch/TableToolbarSearch.jsx:
```diff
@@ -0,0 +1,101 @@
+const TableToolbarSearch = ({
+  defaultValue,
+  defaultExpanded,
+  onChange: onChangeProp,
```
Looks like this is unneeded as the only time `onChange` is used, is when passing it to the `onChange` event handler

enricoberti commented on src/components/Table/TableToolbarSearch/TableToolbarSearch.jsx:
```diff
@@ -0,0 +1,101 @@
+/**
+ * ----------------------------- WARNING ----------------------------
+ * This is a temporary solution to fix a bug in Carbon 9,
+ * to be removed when we import Carbon 10
```
Thanks for this @tay1orjones 
Speaking with @davidicus we cannot just blindly upgrade to 10.10.0 for the MVP downstream consumers so we need to do this fix ourselves

enricoberti commented:
@sstone2423 @tay1orjones  thanks for the reviews,  i've rebased on master and added a couple of tests too to see that the prop aliasing works correctly

stuckless commented on src/components/Table/Table.test.jsx:
```diff
@@ -334,4 +334,48 @@ describe('Table', () => {
+    expect(wrapper.find('.bx--search-input')).toHaveLength(1);
+    expect(wrapper.find('.bx--search-input').prop('value')).toEqual('toyota');
+
+    const wrapper2 = mount(
```
This test is basically the same as the one above it... ie, create a table and verify the value exists (which is a good test, but no need to do it twice).  A better test here for this second one would be use the first wrapper and update the props and then see if the that table instance now reflects the new default search value as set from changing the props.

The use case for how this used is as follows.
We use the table to render the state of a "datasource" which is basically some stateful component that holds things like our data, responds to searches, etc.  That table instance could dynamically load a new "view" which means that it might have a default search, etc, and when that happens we need to the UI to reflect the state of our datasource.  Today, everything about the table, including filters, columns, etc, all comes from our datasource, and it all works woo, with the exception that the "search" value cannot be set in the toolbar.  So we load a new view and show it, and the data is filtered, but, the UI doesn't show that search value.  In essence we have a single table and we just keep updating the props on that table to reflect the state of the data from the datasource.

enricoberti replied:
Gotcha! The duplication was just to make sure aliasing works but, given `value` is being deprecated, I removed it and put something like this in place: https://github.com/IBM/carbon-addons-iot-react/pull/953/commits/f16851c5bbdabe46d120327146f41c11308e5871#diff-f47894e32d4dbf684d2ac3f89df148c9R360

tay1orjones commented:
:tada: This PR is included in version 2.51.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-addons-iot-react)
- [GitHub release](https://github.com/IBM/carbon-addons-iot-react/releases/tag/v2.51.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
